### DISCUSSION TITLE
How does the acknowledgement work for batch consumers?

### DISCUSSION THREAD

**Discussion by yde6010** [spring-projects/spring-kafka#2428] [Q&A]
Can someone explain how the manual commit works when using a batch consumer listener for MANUAL and MANUAL_IMMEDIATE acknowledgement types? For a record listener, this decides if it should commit immediately or commit once all the records consumed during a poll are processed. 

For batch consumers since all the records are sent to the listener what difference will it do with the two options?

The following answer says about being able to catch the exception when using MANUAL with syncCommits. What type of exception is mentioned here? Is it not being able to acknowledge back to the broker?
https://stackoverflow.com/a/60932352/3359134

**Comment by garyrussell** [comment]
>The following answer says about being able to catch the exception when using MANUAL with syncCommits. 

That is incorrect, and not what that answer says.

With `MANUAL`, the commits are queued and processed after the listener exits.

With `MANUAL_IMMEDIATE`, if you are using sync commits, and acknowledge on the listener thread, you will get an exception if the commit fails, allowing you to, for example, log the situation and decide what you want to do in that situation.

**Reply by yde6010** [reply]
Thanks @garyrussell for the explanation. 

Given a batch listener like below, what difference would it have with the two acknowledgement options?

```
@KafkaListener(topics = "myTopic")
public void listen(List<ConsumerRecord<Integer, String>> list, Acknowledgment ack) {
    ...
    ack.acknowledge();
}
```

**Reply by garyrussell** [reply]
Effectively no difference.
